File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex09RevocationReinstatement.cs
using System;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-09 — Revocation with reinstatement.
/// </summary>
/// <remarks>
/// Revocation and reinstatement dates kept separate, and the questions reinstatement does
/// not answer.
/// </remarks>
public sealed class Ex09RevocationReinstatement : IExample
{
    /// <inheritdoc />
    public string Id => "ex-09";

    /// <inheritdoc />
    public string Title => "Revocation and reinstatement dates, kept separate.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        foreach (var ein in new[] { Fixtures.Eins.Revoked, Fixtures.Eins.Reinstated })
        {
            var result = await context.Client.Nonprofits.CheckAsync(ein);
            var organization = result.Nonprofit!;
            var aroe = organization.Aroe();

            Output.Heading(Output.Text(organization.OrganizationName));

            if (aroe is null)
            {
                Output.Bullet("No automatic-revocation data returned.");

                continue;
            }

            Output.DisplayField(aroe, "revocation_date");
            Output.DisplayField(aroe, "reinstatement_date");

            var revokedOn = ApiDate.Parse(aroe.RevocationDate);
            var reinstatedOn = ApiDate.Parse(aroe.ReinstatementDate);

            if (revokedOn is not null && reinstatedOn is not null)
            {
                Output.Field("gap in exemption", $"{(reinstatedOn.Value - revokedOn.Value).Days} days");

                // The gap is the part that matters for a donation made in between: a
                // reinstatement is not always retroactive to the revocation date.
                Output.Bullet(
                    "A donation made inside that window may not be deductible even though the\n"
                    + "    organization is exempt today. Reinstatement is not always retroactive.");
            }

            Output.Field("status today", (revokedOn, reinstatedOn) switch
            {
                (not null, null) => "revoked, not reinstated",
                (not null, not null) => "reinstated after revocation",
                _ => "no revocation on record",
            });
        }

        Output.Heading("What reinstatement does not tell you");
        Output.Bullet("Whether the reinstatement was retroactive to the revocation date.");
        Output.Bullet("Whether donations made during the gap are deductible.");
        Output.Bullet("Why the exemption lapsed in the first place.");

        Output.Note(
            "The API returns two dates. Everything above is a question for the organization\n"
            + "or its counsel — the SDK reports the dates and infers nothing from them.");
    }
}

File: ruby/examples/ex_09_revocation_reinstatement.rb
# frozen_string_literal: true

# EX-09 — Revoked organization with reinstatement data.
#
# A record can carry both a revocation date and a reinstatement date. The two
# stay separately accessible, because the gap between them matters: a donation
# made while the exemption was revoked is not retroactively fixed by a later
# reinstatement, and reinstatement can be retroactive or not.
#
# This example surfaces both dates and the interval, and still routes the record
# to review. Reinstatement resolves one question, not every question.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_09_revocation_reinstatement.rb

require "time"
require_relative "lib/api_date"
require_relative "lib/fixture_api"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

FixtureApi.with_fixture_api do |client|
  result = client.nonprofits.check(FixtureApi::EINS[:reinstated])
  nonprofit = result.nonprofit

  if nonprofit.nil?
    puts "No record returned."
    next
  end

  aroe = NCP::Sources.aroe(nonprofit)

  heading "#{nonprofit.organization_name} (#{nonprofit.ein})"

  # Both dates are their own field. Nothing collapses them into a single
  # "currently revoked" boolean, because that boolean would lose the interval.
  field "revocation_code", returned(aroe, :revocation_code)
  field "revocation_date", returned(aroe, :revocation_date)
  field "reinstatement_date", returned(aroe, :reinstatement_date)

  # The API formats dates as `M/DD/YYYY h:mm:ss AM`. Parse, never reformat in place.
  revoked_at = ApiDate.parse(aroe&.revocation_date)
  reinstated_at = ApiDate.parse(aroe&.reinstatement_date)

  heading "Derived, in application code"

  if revoked_at && reinstated_at
    bullet "revoked on #{revoked_at.strftime("%F")}"
    bullet "reinstated on #{reinstated_at.strftime("%F")}"
    bullet "exemption lapsed for #{ApiDate.age_in_days(revoked_at, reinstated_at)} days"
    bullet "donations dated inside that window may need separate handling"
  elsif revoked_at
    bullet "revoked, with no reinstatement date returned"
  else
    bullet "no revocation history returned"
  end

  heading "What the other sources say now"
  field "bmf_status", returned(NCP::Sources.bmf(nonprofit), :status)
  field "pub78_verified", returned(NCP::Sources.pub78(nonprofit), :verified)
  field "irs_bmf_pub78_conflict", returned(nonprofit, :irs_bmf_pub78_conflict)

  heading "Outcome"

  [
    "Was the reinstatement retroactive to the revocation date?",
    "Do gifts made during the lapse need to be re-characterized?",
    "Does your grant agreement require continuous exemption?",
    "Has the organization filed since reinstatement?"
  ].each { |question| bullet question }

  puts
  field "policy action", "manual review — reinstatement is recorded, and the record still has history"

  puts "\n  evidence retained:"
  json_block(
    { ein: nonprofit.ein, request_id: result.request_id, checked_at: Time.now.utc.iso8601 }
      .merge(nonprofit.to_h.slice("revocation_date", "reinstatement_date", "revocation_code"))
  )
end

note "The API answers \"what does the IRS revocation data show\". It does not answer\n" \
     "\"is this organization eligible today\" — that needs your policy, and often your\n" \
     "counsel."
